pilot/pkg/xds/eds.go

```diff
@@ -160,61 +160,74 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
 	}
 
 	ep.mutex.Lock()
-	oldIstioEndpoints := ep.Shards[shard]
-	ep.Shards[shard] = istioEndpoints
-	// Check if ServiceAccounts have changed. We should do a full push if they have changed.
-	saUpdated := s.UpdateServiceAccount(ep, hostname)
+	defer ep.mutex.UnLock()
+	newIstioEndpoints := []*model.IstioEndpoint{}
```

Comment:
```suggestion
	newIstioEndpoints := oldIstioEndpoints
	if features.SendUnhealthyEndpoints {
	   newIstioEndpoints = make([]*model.IstioEndpoint, 0, len(oldIstioEndpoints))
	   
	   
```

Benefits are then on L203 we do not need the switch cases, we just always set to `newIstioEndpoints`. But the bigger thing is pre-allocating the array, since this can be 100s of endpoints and grow a lot